examples/microphone_nextjs/utils/recorder.ts

```diff
@@ -73,86 +73,125 @@ export class AudioRecorder {
+    const devices: MediaDeviceInfo[] =
+      await navigator.mediaDevices.enumerateDevices();
+    const filtered = devices.filter((device: MediaDeviceInfo) => {
+      return (
+        device.kind === 'audioinput' &&
+        device.deviceId !== '' &&
+        device.label !== ''
+      );
+    });
+    this.devices = filtered;
+  };
 
-  return [devices, denied];
-}
+  private promptAudioInputs = async () => {
+    const permissions = await getPermissions();
+    if (permissions === 'denied') {
+      this.denied = true;
+      return;
+    }
 
-// getAudioInputs enumerates all the available microphones
-async function getAudioInputs(): Promise<MediaDeviceInfo[]> {
-  // We start by checking permissions
-  // If permissions are denied, throw an error
-  if ((await getPermissions()) === 'denied')
-    throw new Error('Permissions denied');
+    // If permissions are prompt, we need to call getUserMedia to ask the user for permission
+    if (permissions === 'prompt') {
+      await this.getUserMedia({
+        audio: true,
+        video: false,
+      });
+    } else {
+      this.updateDeviceList();
+    }
+  };
+}
+const audioDevices = new AudioDevices();
 
-  // If permissions are prompt, we need to call getUserMedia to ask the user for permission
-  if ((await getPermissions()) === 'prompt') {
-    await navigator.mediaDevices.getUserMedia({ audio: true, video: false });
-  }
+// Devices state
+function subscribeDevices(callback) {
+  audioDevices.addEventListener('changeDevices', callback);
+  return () => {
+    audioDevices.removeEventListener('changeDevices', callback);
+  };
+}
+const getDevices = () => audioDevices.devices;
+export function useAudioDevices() {
+  return useSyncExternalStore(subscribeDevices, getDevices, getDevices);
+}
 
-  // now we have permissions, we attempt to get the audio devices
-  const devices: MediaDeviceInfo[] =
-    await navigator.mediaDevices.enumerateDevices();
-  const filtered = devices.filter((device: MediaDeviceInfo) => {
-    return device.kind === 'audioinput';
-  });
-  // If labels are null, try opening streams to get labels
-  // This is only for firefox where the device label can only be read when permission
-  // has been given to access each device, not just audio devices in general
-  if (!filtered[0].label) {
-    return await getAudioInputsOpenStreams();
-  }
-  return filtered;
+// Denied state
+function subscribeDenied(callback) {
+  audioDevices.addEventListener('changeDenied', callback);
+  return () => {
+    audioDevices.removeEventListener('changeDenied', callback);
+  };
+}
+const getDenied = () => audioDevices.denied;
+export function useAudioDenied() {
+  return useSyncExternalStore(subscribeDenied, getDenied, getDenied);
 }
 
-// This function opens streams as Firefox only allows read access to open streams
-// We do this just to populate the streams list and then close them
-async function getAudioInputsOpenStreams(): Promise<MediaDeviceInfo[]> {
```

Comment:
@TudorCRL after making my changes and testing across chromium (116), Safari (16.6 macos) and Firefox (117.0.1), I could no longer find a need for this step so I have removed it. Perhaps I'm missing something - may be good for you to check out my branch and try for yourself?

Reply:
Yeah sure I'll have a look

Reply:
Looks good to me